src/MOI_wrapper.jl

```diff
@@ -0,0 +1,183 @@
+import SolverCore
+
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    options::Dict{String,Any}
+    silent::Bool
+    solver
+    nlp::Union{Nothing,MathOptNLPModel}
+    stats::Union{
+        Nothing,
+        SolverCore.GenericExecutionStats{Float64,Vector{Float64},Vector{Float64},Any},
+    }
+    function Optimizer()
+        return new(Dict{String,Any}(), false, nothing, nothing, nothing)
+    end
+end
+
+# FIXME return the name of the underlying NLPModel solver
+MOI.get(::Optimizer, ::MOI.SolverName) = "NLPModels"
+
+MOI.is_empty(optimizer::Optimizer) = isnothing(optimizer.solver) && isnothing(optimizer.nlp)
+
+function MOI.empty!(optimizer::Optimizer)
+    optimizer.solver = nothing
+    optimizer.nlp = nothing
+    optimizer.stats = nothing
+    return
+end
+
+###
+### MOI.RawOptimizerAttribute
+###
+
+function MOI.set(optimizer::Optimizer, param::MOI.RawOptimizerAttribute, value)
+    return optimizer.options[param.name] = value
+end
+
+function MOI.get(optimizer::Optimizer, param::MOI.RawOptimizerAttribute)
+    return optimizer.options[param.name]
+end
+
+###
+### MOI.Silent
+###
+
+MOI.supports(::Optimizer, ::MOI.Silent) = true
+
+function MOI.set(optimizer::Optimizer, ::MOI.Silent, value::Bool)
+    optimizer.silent = value
+    return
+end
+
+MOI.get(optimizer::Optimizer, ::MOI.Silent) = optimizer.silent
+
+###
+### MOI.AbstractModelAttribute
+###
+
+function MOI.supports(
+    ::Optimizer,
+    ::Union{
+        MOI.ObjectiveSense,
+        MOI.ObjectiveFunction{<:OBJ},
+        MOI.NLPBlock,
+    },
+)
+    return true
+end
+
+###
+### MOI.AbstractVariableAttribute
+###
+
+function MOI.supports(
+    ::Optimizer,
+    ::MOI.VariablePrimalStart,
+    ::Type{MOI.VariableIndex},
+)
+    return true
+end
+
+###
+### `supports_constraint`
+###
+
+MOI.supports_constraint(::Optimizer, ::Type{VI}, ::Type{<:ALS}) = true
+MOI.supports_constraint(::Optimizer, ::Type{SAF}, ::Type{<:ALS}) = true
+MOI.supports_constraint(::Optimizer, ::Type{VAF}, ::Type{<:VLS}) = true
+
+function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike)
+    if !haskey(dest.options, "solver")
+        error("No solver specified, use for instance `using Percival; JuMP.set_attribute(model, \"solver\", PercivalSolver)`")
+    end
+    dest.nlp, index_map = nlp_model(src)
+    dest.solver = dest.options["solver"](dest.nlp)
+    return index_map
+end
+
+function MOI.optimize!(model::Optimizer)
+    options = Dict{Symbol,Any}(
+        Symbol(key) => model.options[key]
+        for key in keys(model.options) if key != "solver"
+    )
+    if model.silent
+        options[:verbose] = 0
+    end
+    model.stats = SolverCore.solve!(model.solver, model.nlp; options...)
```

Comment:
About updating stats in a hot loop, you should use this:
```suggestion
    SolverCore.solve!(model.solver, model.nlp, model.stats; options...)
```